### PR TITLE
Issue #1724: Strip leading parts of paths in compiled code objects

### PR DESCRIPTION
**Pull request by codewarrior0**
Reverts the previous implementation. Implements path stripping in PYZ immediately before calling ZlibArchiveWriter. Path stripping code adapted from ModuleGraph's `_replace_paths_in_code` method, which seemed to do the job the first time around. (See also test case in 0a70a1bda192fc7f051b02198f36349d53e21daf)

The `strip_paths` option is no more! Now, the entire path found in `sys.path + pathex` is always stripped from filenames.

Submitted for @htgoebel's approval.

### CODE REVIEW

File: PyInstaller/building/api.py [hunk 1]
```diff
@@ -171,8 +177,44 @@ def assemble(self):
                 self.code_dict[entry[0]] = self.__get_code(entry[0], entry[1])
         # sort content alphabetically to support reproducible builds
         toc.sort()
+
+        # Remove leading parts of paths in code objects
+        self.code_dict = {
+            key: self._strip_paths_in_code(code)
+            for key, code in self.code_dict.items()
+        }
```
Comment (htgoebel) [inline PyInstaller/building/api.py]: Wow, this is interesting. I've never seen this before. I used to write `dict((key, value) for ...)`. But this is much nicer :-)

**Comment by htgoebel** [comment]
@codewarrior0 Thanks a lot taking the time working in this again.

This is okay for me new.Thanks again.

If you'd like to improve the code taken from modulegraph a bit: I've added some line-comments.

File: PyInstaller/building/api.py [hunk 2]
```diff
@@ -171,8 +177,44 @@ def assemble(self):
...
+
         pyz = ZlibArchiveWriter(self.name, toc, code_dict=self.code_dict, cipher=self.cipher)
 
+    def _strip_paths_in_code(self, co):
```
Comment (htgoebel) [inline PyInstaller/building/api.py]: If we'd add another parameter `new_filename=None` we'd only need to calculate the new filename once.

File: PyInstaller/building/api.py [hunk 3]
```diff
@@ -171,8 +177,44 @@ def assemble(self):
...
+        for f in self.replace_paths:
+            if original_filename.startswith(f):
+                new_filename = original_filename[len(f):]
+                break
+
+        else:
+            return co
+
+        consts = list(co.co_consts)
+        for i in range(len(consts)):
```
Comment (htgoebel) [inline PyInstaller/building/api.py]: We could use `enumerate` here, as we only support 2.7 only. We could even use list comprehensions here.

File: PyInstaller/building/api.py (hunk 1 above)
Comment (codewarrior0) [inline PyInstaller/building/api.py]: Yep, it's called a "dict comprehension", introduced in 2.7.

File: PyInstaller/building/api.py (hunk 2 above)
Comment (codewarrior0) [inline PyInstaller/building/api.py]: Good point. That's better than recalculating it for each recursive call.

**Comment by codewarrior0** [comment]
Done.

**Comment by htgoebel** [comment]
@codewarrior0 Thanks for hunting down this long-running issue :-)

**Pull request merged**